File: phases/04-algorithms-and-complexity/22-computational-geometry-i-convex-hull-sweep-line/code/main.cpp
// Computational Geometry I — Convex Hull, Sweep Line
// Phase 04 — Algorithms & Complexity Analysis
//
// C++ geometry toolkit: cross product, Graham scan, Jarvis march,
// closest pair sweep line, segment intersection.
#include <algorithm>
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <iostream>
#include <random>
#include <vector>

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

struct Point {
    double x, y;
    bool operator<(const Point& o) const {
        return std::tie(x, y) < std::tie(o.x, o.y);
    }
    bool operator==(const Point& o) const {
        return x == o.x && y == o.y;
    }
};

// ---------------------------------------------------------------------------
// Primitives
// ---------------------------------------------------------------------------

// Cross product of vectors o->a and o->b.
// Positive => CCW, Negative => CW, Zero => collinear.
static double cross(const Point& o, const Point& a, const Point& b) {
    return (a.x - o.x) * (b.y - o.y) - (a.y - o.y) * (b.x - o.x);
}

static double dist2(const Point& a, const Point& b) {
    double dx = a.x - b.x, dy = a.y - b.y;
    return dx * dx + dy * dy;
}

static double dist(const Point& a, const Point& b) {
    return std::sqrt(dist2(a, b));
}

// ---------------------------------------------------------------------------
// Convex Hull: Graham Scan  O(n log n)
// ---------------------------------------------------------------------------

std::vector<Point> graham_scan(std::vector<Point> pts) {
    if (pts.size() <= 2) return pts;

    // Step 1: lowest point (y, then x)
    int idx = 0;
    for (int i = 1; i < (int)pts.size(); ++i) {
        if (pts[i].y < pts[idx].y ||
            (pts[i].y == pts[idx].y && pts[i].x < pts[idx].x))
            idx = i;
    }
    std::swap(pts[0], pts[idx]);
    Point start = pts[0];

    // Step 2: sort by polar angle from start
    std::sort(pts.begin() + 1, pts.end(),
              [&start](const Point& a, const Point& b) {
                  double ang_a = std::atan2(a.y - start.y, a.x - start.x);
                  double ang_b = std::atan2(b.y - start.y, b.x - start.x);
                  return ang_a < ang_b;
              });

    // Step 3: build hull with stack
    std::vector<Point> hull;
    for (auto& p : pts) {
        while (hull.size() >= 2 &&
               cross(hull[hull.size() - 2], hull.back(), p) <= 0)
            hull.pop_back();
        hull.push_back(p);
    }
    return hull;
}

// ---------------------------------------------------------------------------
// Convex Hull: Jarvis March (Gift Wrapping)  O(n h)
// ---------------------------------------------------------------------------

std::vector<Point> jarvis_march(const std::vector<Point>& pts) {
    if (pts.size() <= 2) return pts;

    // Leftmost point
    int start_idx = 0;
    for (int i = 1; i < (int)pts.size(); ++i) {
        if (pts[i].x < pts[start_idx].x ||
            (pts[i].x == pts[start_idx].x && pts[i].y < pts[start_idx].y))
            start_idx = i;
    }

    std::vector<Point> hull;
    int current = start_idx;

    while (true) {
        hull.push_back(pts[current]);
        int candidate = (current != 0) ? 0 : 1;

        for (int i = 0; i < (int)pts.size(); ++i) {
            if (i == current) continue;
            double cp = cross(pts[current], pts[candidate], pts[i]);
            if (cp < 0) {
                candidate = i;
            } else if (cp == 0) {
                if (dist2(pts[current], pts[i]) > dist2(pts[current], pts[candidate]))
                    candidate = i;
            }
        }

        current = candidate;
        if (current == start_idx) break;
    }
    return hull;
}

// ---------------------------------------------------------------------------
// Closest Pair (Sweep Line)  O(n log n)
// ---------------------------------------------------------------------------

struct ClosestResult {
    double dist;
    Point a, b;
    bool valid;
};

ClosestResult closest_pair_sweep(std::vector<Point> pts) {
    ClosestResult res{1e18, {}, {}, false};
    if (pts.size() < 2) return res;

    std::sort(pts.begin(), pts.end());
    std::vector<Point> active;

    int j = 0;
    for (auto& p : pts) {
        while (j < (int)pts.size() && pts[j].x < p.x - res.dist)
            ++j;

        // Build strip from active points within vertical distance < best
        active.clear();
        for (int k = j; k < (int)pts.size() && pts[k].x <= p.x + res.dist; ++k) {
            if (pts[k] == p) continue;
            if (std::abs(pts[k].y - p.y) < res.dist)
                active.push_back(pts[k]);
        }
        std::sort(active.begin(), active.end(),
                  [](const Point& a, const Point& b) { return a.y < b.y; });

        for (auto& q : active) {
            double d = dist(p, q);
            if (d < res.dist) {
                res.dist = d;
                res.a = p;
                res.b = q;
                res.valid = true;
            }
        }
    }
    return res;
}

// ---------------------------------------------------------------------------
// Segment Intersection
// ---------------------------------------------------------------------------

static bool on_segment(const Point& p, const Point& q, const Point& r) {
    return (std::min(p.x, q.x) <= r.x && r.x <= std::max(p.x, q.x) &&
            std::min(p.y, q.y) <= r.y && r.y <= std::max(p.y, q.y));
}

bool segments_intersect(const Point& p1, const Point& p2,
                        const Point& p3, const Point& p4) {
    double d1 = cross(p3, p4, p1);
    double d2 = cross(p3, p4, p2);
    double d3 = cross(p1, p2, p3);
    double d4 = cross(p1, p2, p4);

    if (((d1 > 0 && d2 < 0) || (d1 < 0 && d2 > 0)) &&
        ((d3 > 0 && d4 < 0) || (d3 < 0 && d4 > 0)))
        return true;

    if (d1 == 0 && on_segment(p3, p4, p1)) return true;
    if (d2 == 0 && on_segment(p3, p4, p2)) return true;
    if (d3 == 0 && on_segment(p1, p2, p3)) return true;
    if (d4 == 0 && on_segment(p1, p2, p4)) return true;
    return false;
}

// ---------------------------------------------------------------------------
// Hull Area (Shoelace Formula)
// ---------------------------------------------------------------------------

double hull_area(const std::vector<Point>& hull) {
    double area = 0.0;
    int n = (int)hull.size();
    for (int i = 0; i < n; ++i) {
        int j = (i + 1) % n;
        area += hull[i].x * hull[j].y;
        area -= hull[j].x * hull[i].y;
    }
    return std::abs(area) / 2.0;
}

// ---------------------------------------------------------------------------
// Demo
// ---------------------------------------------------------------------------

int main() {
    std::mt19937 rng(42);
    std::uniform_real_distribution<double> dist_rand(0.0, 99.0);

    const int N = 40;
    std::vector<Point> points(N);
    for (auto& p : points) {
        p.x = dist_rand(rng);
        p.y = dist_rand(rng);
    }

    std::cout << "==========================================\n";
    std::cout << "  Computational Geometry I — C++ Toolkit\n";
    std::cout << "==========================================\n\n";

    // --- Graham Scan ---
    auto hull_graham = graham_scan(points);
    std::cout << "--- Graham Scan ---\n";
    std::cout << "Points: " << points.size()
              << "  Hull vertices: " << hull_graham.size() << "\n";

    // --- Jarvis March ---
    auto hull_jarvis = jarvis_march(points);
    std::cout << "\n--- Jarvis March ---\n";
    std::cout << "Points: " << points.size()
              << "  Hull vertices: " << hull_jarvis.size() << "\n";

    // --- Orientation demo ---
    std::cout << "\n--- Orientation Test ---\n";
    Point o{0, 0}, a{1, 0}, b_ccw{1, 1}, b_cw{1, -1}, b_col{2, 0};
    std::cout << "cross (0,0)->(1,0)->(1,1)  = " << cross(o, a, b_ccw) << "  (CCW)\n";
    std::cout << "cross (0,0)->(1,0)->(1,-1) = " << cross(o, a, b_cw) << "  (CW)\n";
    std::cout << "cross (0,0)->(1,0)->(2,0)  = " << cross(o, a, b_col) << "  (collinear)\n";

    // --- Closest Pair ---
    auto cp = closest_pair_sweep(points);
    std::cout << "\n--- Closest Pair (Sweep Line) ---\n";
    if (cp.valid) {
        std::cout << "Distance: " << cp.dist << "\n";
        std::cout << "Points: (" << cp.a.x << "," << cp.a.y
                  << ") and (" << cp.b.x << "," << cp.b.y << ")\n";
    }

    // --- Segment Intersection ---
    std::cout << "\n--- Segment Intersection ---\n";
    std::cout << "(0,0)-(4,4) vs (0,4)-(4,0): "
              << segments_intersect({0, 0}, {4, 4}, {0, 4}, {4, 0}) << "\n";
    std::cout << "(0,0)-(2,2) vs (3,3)-(5,5): "
              << segments_intersect({0, 0}, {2, 2}, {3, 3}, {5, 5}) << "\n";
    std::cout << "(0,0)-(4,4) vs (3,3)-(5,5): "
              << segments_intersect({0, 0}, {4, 4}, {3, 3}, {5, 5}) << "\n";

    // --- Hull Area ---
    double area = hull_area(hull_graham);
    std::cout << "\n--- Hull Area (Shoelace) ---\n";
    std::cout << "Convex hull area: " << area << "\n";

    return 0;
}

File: phases/04-algorithms-and-complexity/22-computational-geometry-i-convex-hull-sweep-line/code/main.py
"""
Computational Geometry I — Convex Hull, Sweep Line
Phase 04 — Algorithms & Complexity Analysis

Implementations:
  - cross product / orientation test
  - Graham scan  O(n log n)
  - Jarvis march  O(n h)
  - Closest pair (sweep line)  O(n log n)
  - Segment intersection detection
"""

import math
from typing import List, Tuple, Optional

Point = Tuple[float, float]
Segment = Tuple[Point, Point]


# ---------------------------------------------------------------------------
# Primitives
# ---------------------------------------------------------------------------

def cross(o: Point, a: Point, b: Point) -> float:
    """Cross product of vectors o->a and o->b.
    Positive => CCW, Negative => CW, Zero => collinear.
    """
    return (a[0] - o[0]) * (b[1] - o[1]) - (a[1] - o[1]) * (b[0] - o[0])


def dist2(a: Point, b: Point) -> float:
    """Squared Euclidean distance."""
    return (a[0] - b[0]) ** 2 + (a[1] - b[1]) ** 2


def dist(a: Point, b: Point) -> float:
    """Euclidean distance."""
    return math.sqrt(dist2(a, b))


# ---------------------------------------------------------------------------
# ASCII visualisation
# ---------------------------------------------------------------------------

def ascii_plot(points: List[Point], hull: Optional[List[Point]] = None,
               width: int = 60, height: int = 20) -> str:
    """Render points and optional hull in an ASCII grid."""
    if not points:
        return "(no points)"

    xs = [p[0] for p in points]
    ys = [p[1] for p in points]
    min_x, max_x = min(xs), max(xs)
    min_y, max_y = min(ys), max(ys)

    # Guard against degenerate range
    if max_x == min_x:
        max_x = min_x + 1
    if max_y == min_y:
        max_y = min_y + 1

    grid = [[' ' for _ in range(width)] for _ in range(height)]

    def to_grid(px, py):
        gx = int((px - min_x) / (max_x - min_x) * (width - 1))
        gy = int((1 - (py - min_y) / (max_y - min_y)) * (height - 1))
        gx = max(0, min(width - 1, gx))
        gy = max(0, min(height - 1, gy))
        return gx, gy

    # Draw hull edges as '.'
    if hull and len(hull) >= 2:
        for i in range(len(hull)):
            x0, y0 = to_grid(hull[i][0], hull[i][1])
            x1, y1 = to_grid(hull[(i + 1) % len(hull)][0], hull[(i + 1) % len(hull)][1])
            steps = max(abs(x1 - x0), abs(y1 - y0), 1)
            for s in range(steps + 1):
                cx = x0 + int(round((x1 - x0) * s / steps))
                cy = y0 + int(round((y1 - y0) * s / steps))
                if grid[cy][cx] == ' ':
                    grid[cy][cx] = '.'

    # Draw hull vertices as '#'
    if hull:
        for hx, hy in hull:
            gx, gy = to_grid(hx, hy)
            grid[gy][gx] = '#'

    # Draw interior points as '*'
    for px, py in points:
        gx, gy = to_grid(px, py)
        if grid[gy][gx] == ' ':
            grid[gy][gx] = '*'

    border = '+' + '-' * width + '+'
    rows = [border]
    for row in grid:
        rows.append('|' + ''.join(row) + '|')
    rows.append(border)
    return '\n'.join(rows)


# ---------------------------------------------------------------------------
# Convex Hull: Graham Scan
# ---------------------------------------------------------------------------

def graham_scan(points: List[Point]) -> List[Point]:
    """Convex hull via Graham scan. Returns hull vertices in CCW order.

    Complexity: O(n log n) — dominated by the sort.
    """
    if len(points) <= 2:
        return list(points)

    # Step 1: find lowest point (lowest y, then leftmost x)
    start = min(points, key=lambda p: (p[1], p[0]))

    # Step 2: sort by polar angle from start
    def polar_key(p):
        return math.atan2(p[1] - start[1], p[0] - start[0])

    sorted_pts = sorted(points, key=polar_key)

    # Step 3: build hull — pop on right turns (cross <= 0)
    hull: List[Point] = []
    for p in sorted_pts:
        while len(hull) >= 2 and cross(hull[-2], hull[-1], p) <= 0:
            hull.pop()
        hull.append(p)

    return hull


# ---------------------------------------------------------------------------
# Convex Hull: Jarvis March (Gift Wrapping)
# ---------------------------------------------------------------------------

def jarvis_march(points: List[Point]) -> List[Point]:
    """Convex hull via gift wrapping. Returns hull vertices in CCW order.

    Complexity: O(n h) where h = hull size.
    """
    if len(points) <= 2:
        return list(points)

    start = min(points, key=lambda p: (p[0], p[1]))
    hull: List[Point] = []
    current = start

    while True:
        hull.append(current)
        # Pick a candidate that is not current
        candidate = points[0] if points[0] != current else points[1]

        for p in points:
            if p == current:
                continue
            cp = cross(current, candidate, p)
            if cp < 0:
                candidate = p
            elif cp == 0:
                # Collinear — take the farther one
                if dist2(current, p) > dist2(current, candidate):
                    candidate = p

        current = candidate
        if current == start:
            break

    return hull


# ---------------------------------------------------------------------------
# Closest Pair (Sweep Line)
# ---------------------------------------------------------------------------

def closest_pair_sweep(points: List[Point]) -> Tuple[float, Optional[Tuple[Point, Point]]]:
    """Find closest pair of points using sweep line.

    Complexity: O(n log n) — sort + active-set maintenance.

    Returns (distance, (point_a, point_b)).
    """
    if len(points) < 2:
        return float('inf'), None

    pts = sorted(points, key=lambda p: p[0])  # sort by x
    best = float('inf')
    best_pair: Optional[Tuple[Point, Point]] = None
    active: List[Point] = []

    j = 0
    for p in pts:
        # Evict points whose x is more than best behind the sweep line
        while j < len(pts) and pts[j][0] < p[0] - best:
            if pts[j] in active:
                active.remove(pts[j])
            j += 1

        # Build strip: active points within vertical distance < best
        strip = [q for q in active if abs(q[1] - p[1]) < best]
        strip.sort(key=lambda q: q[1])

        for q in strip:
            d = dist(p, q)
            if d < best:
                best = d
                best_pair = (p, q)

        active.append(p)

    return best, best_pair


# ---------------------------------------------------------------------------
# Segment Intersection
# ---------------------------------------------------------------------------

def _on_segment(p: Point, q: Point, r: Point) -> bool:
    """Is r on segment p-q (assuming collinear)?"""
    return (min(p[0], q[0]) <= r[0] <= max(p[0], q[0]) and
            min(p[1], q[1]) <= r[1] <= max(p[1], q[1]))


def segments_intersect(s1: Segment, s2: Segment) -> bool:
    """Check whether two line segments (p1,p2) and (p3,p4) intersect."""
    p1, p2 = s1
    p3, p4 = s2

    d1 = cross(p3, p4, p1)
    d2 = cross(p3, p4, p2)
    d3 = cross(p1, p2, p3)
    d4 = cross(p1, p2, p4)

    if ((d1 > 0 and d2 < 0) or (d1 < 0 and d2 > 0)) and \
       ((d3 > 0 and d4 < 0) or (d3 < 0 and d4 > 0)):
        return True

    # Collinear endpoint cases
    if d1 == 0 and _on_segment(p3, p4, p1):
        return True
    if d2 == 0 and _on_segment(p3, p4, p2):
        return True
    if d3 == 0 and _on_segment(p1, p2, p3):
        return True
    if d4 == 0 and _on_segment(p1, p2, p4):
        return True

    return False


# ---------------------------------------------------------------------------
# Demo
# ---------------------------------------------------------------------------

def main() -> None:
    import random
    random.seed(42)

    points = [(random.randint(0, 99), random.randint(0, 99)) for _ in range(40)]

    print("=" * 62)
    print("  Computational Geometry I — Convex Hull, Sweep Line")
    print("=" * 62)

    # --- Graham Scan ---
    hull_graham = graham_scan(points)
    print(f"\n--- Graham Scan ---")
    print(f"Points: {len(points)},  Hull vertices: {len(hull_graham)}")
    print(ascii_plot(points, hull_graham))

    # --- Jarvis March ---
    hull_jarvis = jarvis_march(points)
    print(f"\n--- Jarvis March ---")
    print(f"Points: {len(points)},  Hull vertices: {len(hull_jarvis)}")

    # Verify both hulls produce the same set of vertices
    set_graham = set(hull_graham)
    set_jarvis = set(hull_jarvis)
    print(f"Hull match: {set_graham == set_jarvis}")

    # --- Orientation demo ---
    print("\n--- Orientation Test ---")
    o, a, b = (0, 0), (1, 0), (1, 1)
    print(f"cross{o}->{a}->{b} = {cross(o, a, b):.1f}  (CCW)")
    o, a, b = (0, 0), (1, 0), (1, -1)
    print(f"cross{o}->{a}->{b} = {cross(o, a, b):.1f}  (CW)")
    o, a, b = (0, 0), (1, 0), (2, 0)
    print(f"cross{o}->{a}->{b} = {cross(o, a, b):.1f}  (collinear)")

    # --- Closest Pair ---
    d, pair = closest_pair_sweep(points)
    print(f"\n--- Closest Pair (Sweep Line) ---")
    print(f"Distance: {d:.4f}")
    if pair:
        print(f"Points: {pair[0]} and {pair[1]}")

    # --- Segment Intersection ---
    print("\n--- Segment Intersection ---")
    s1 = ((0, 0), (4, 4))
    s2 = ((0, 4), (4, 0))
    s3 = ((0, 0), (2, 2))
    s4 = ((3, 3), (5, 5))
    print(f"Segments (0,0)-(4,4) and (0,4)-(4,0): {segments_intersect(s1, s2)}")
    print(f"Segments (0,0)-(2,2) and (3,3)-(5,5): {segments_intersect(s3, s4)}")
    print(f"Segments (0,0)-(4,4) and (3,3)-(5,5): {segments_intersect(s1, s4)}")

    # --- Hull area (shoelace formula) ---
    def hull_area(hull: List[Point]) -> float:
        n = len(hull)
        area = 0.0
        for i in range(n):
            j = (i + 1) % n
            area += hull[i][0] * hull[j][1]
            area -= hull[j][0] * hull[i][1]
        return abs(area) / 2.0

    area = hull_area(hull_graham)
    print(f"\n--- Hull Area (Shoelace) ---")
    print(f"Convex hull area: {area:.2f}")


if __name__ == "__main__":
    main()
